apps/remote_control/test/lexical/remote_control/search/indexer/extractors/variable_test.exs

```diff
@@ -0,0 +1,440 @@
+defmodule Lexical.RemoteControl.Search.Indexer.Extractors.VariableTest do
+  alias Lexical.RemoteControl.Search.Indexer.Extractors
+
+  use Lexical.Test.ExtractorCase
+
+  def index_references(source) do
+    do_index(source, fn entry -> entry.type == :variable and entry.subtype == :reference end, [
+      Extractors.Variable
+    ])
+  end
+
+  def index_definitions(source) do
+    do_index(source, fn entry -> entry.type == :variable and entry.subtype == :definition end, [
+      Extractors.Variable
+    ])
+  end
+
+  def assert_definition(entry, variable_name) do
+    assert entry.type == :variable
+    assert entry.subtype == :definition
+    assert entry.subject == variable_name
+  end
+
+  def assert_reference(entry, variable_name) do
+    assert entry.type == :variable
+    assert entry.subtype == :reference
+    assert entry.subject == variable_name
+  end
+
+  for def_type <- [:def, :defp, :defmacro, :defmacrop] do
```

Comment:
I like the simplicity of this PR, but I'm curious, besides the unimplemented cross-scope variable extraction, what other scenarios have we deliberately not implemented?

Regarding testing, I think we're missing cases where definitions and references are extracted together, like this:

```elixir
  test "definition at `=`'s right side" do
    {:ok, [ref], doc} = ~q/
        fn
          1 = a -> a
        end
      / |> index_references()

    assert decorate(doc, ref.range) =~ "1 = a -> «a»"
  end

  test "simple reference" do
    {:ok, [ref], doc} = ~q/
      def myfunc(a) do
        [a]
      end
      / |> index_references()

    assert decorate(doc, ref.range) =~ "[«a»]"
  end
```

**Note**, the current code cannot pass these two tests.

Additionally, how would you differentiate between multiple variables with the same name appearing in a single scope?

Reply:
@scottming, the first example isn't valid syntax, so it's going to fail and produce no results. The second is a valid bug though.

Reply:
I don't think the first example is incorrect syntax; it can be run directly in iex. Also, with a simple function like this: `fn x -> x end`, we can't extract the variables and references. I think we need more tests with references and definitions together.

![CleanShot 2024-02-28 at 20 36 38@2x](https://github.com/lexical-lsp/lexical/assets/12830256/905fb8dd-7505-4aba-b8ba-c3704f603e58)

Reply:
The syntax you typed was not inside an anonymous function, and that works, even in the old code.

Reply:
You're right, I missed the lone `fn` on the line above.